### DISCUSSION TITLE
Pass a mutable reference

### DISCUSSION THREAD

**Discussion by therealevahill** [PyO3/pyo3#4230] [Questions] [answered]
Hi,

at first, thank your great work. It is really a pleasure to work with PyO3.

I have one question, I did not find an answer for it.
How can I pass a mutalbe reference u16 from a python function call into PyO3 rust?

```rust
#[pyfunction]
fn foo(value: &mut u16) {
    *value = 42;
}
```
```python
value = 0
foo(value)
print(value)  # would expect to print 42 here.
```

I tried lots of stuff with PyCell and Bound, but did not get it working.

**Comment by alex** [comment]
You can't. Integers are immutable in Python, and Python does not support passing reference-to-variables in the way your example shows.

**Reply by therealevahill** [reply]
Is it then possible with a struct?

**Reply by alex** [reply]
Yes, it's possible to receive a mutable reference to a `pyclass` struct: https://pyo3.rs/v0.21.2/class#bound-and-interior-mutability

**Reply by therealevahill** [reply]
You said, that it is not possible in the way the example shows. What are the other possibilities?
With ctypes it works like this, if calling a dll/so.

```python
pointer_to_u16 = ctypes.pointer(ctypes.c_uint16(0))
foo(pointer_to_u16)
content = pointer_to_u16.contents.value
```

**Reply by therealevahill** [reply, marked as answer]
> Yes, it's possible to receive a mutable reference to a `pyclass` struct: https://pyo3.rs/v0.21.2/class#bound-and-interior-mutability

Doing like this works: Thank you very much @alex 

```rust
use pyo3::prelude::*;

#[pyclass]
struct MyClass {
    pub inner: i32,
}

#[pymethods]
impl MyClass {
    #[new]
    fn new(value: i32) -> Self {
        MyClass { inner: value }
    }

    fn show(&self) {
        println!("MyClass: {}", self.inner);
    }
}


#[pyfunction]
fn adapt(x: &mut MyClass) {
    x.inner = 42;
}

#[pymodule]
fn foo_rs(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_class::<MyClass>()?;
    m.add_function(wrap_pyfunction!(adapt, m)?)?;
    Ok(())
}
```
```python
from foo_rs import adapt, MyClass

c = MyClass(37)
c.show()
adapt(c)
c.show()
```

**Reply by therealevahill** [reply]
@alex Could you be so kind, and tell me, what the other possibilies would look like?

You said, that it is not possible in the way the example shows. What are the other possibilities?
With ctypes it works like this, if calling a dll/so.

pointer_to_u16 = ctypes.pointer(ctypes.c_uint16(0))
foo(pointer_to_u16)
content = pointer_to_u16.contents.value
